Error on clean install - Alpha 5

Log file attached.

Error:
```
Error: ENOENT: no such file or directory, stat 'C:\Users\Geras\AppData\Roaming\exile-diary\settings.json'
```
OS: Windows NT 10.0.22621
[new.txt](https://github.com/Qt-dev/exile-diary/files/11423728/new.txt)

Found the issue. I got the fix, I will probably release it aigu the API changes, which may also take care of the errors like that 429 you have been getting
On May 8, 2023 at 9:13 PM -0700, mSoul01 ***@***.***>, wrote:
> Yeah. I can't get it to start unless I install the original version of ED. It's looking in that folder for the settings.json file.
> [2023-05-08 13:46:29.734] [error] (main) Unhandled rejection Error: Cannot find module 'C:\Users\Geras\AppData\Roaming\exile-diary\settings.json'
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you commented.Message ID: ***@***.***>

--- a/src/main/AuthManager.ts
+++ b/src/main/AuthManager.ts
@@ -97,11 +97,13 @@ const AuthManager = {
     const password = await keytar.getPassword(service, account);
     const expirationDate = SettingsManager.get('tokenExpirationDate');
     const username = SettingsManager.get('username');
+    const activeProfile = SettingsManager.get('activeProfile');
     const isAuthenticated =
       password !== null &&
       expirationDate !== null &&
       moment().isBefore(expirationDate) &&
-      !!username;
+      !!username &&
+      !!activeProfile && !!activeProfile.characterName && !!activeProfile.league && !!activeProfile.valid;
     logger.info(`User is ${isAuthenticated ? '' : 'not '}authenticated`, {
       password: !!password,
       expirationDate,

--- a/src/main/GGGAPI.ts
+++ b/src/main/GGGAPI.ts
@@ -198,9 +198,9 @@ const APIManager = {
     );
     return currentCharacter;
   },
-  getAllCharacters: async () => {
+  getAllCharacters: async () : Promise<any[]> => {
     const characters = await getAllCharacters();
-    return characters;
+    return characters ?? [];
   },
   getDataForInventory: async () => {
     const inventory = await getDataForInventory();

--- a/src/main/RatesManager.ts
+++ b/src/main/RatesManager.ts
@@ -19,6 +19,9 @@ class RatesManager {
     if (!this.rates[date] || !this.rates[date][league]) {
       await this.fetchRatesForDay(league, date);
     }
+    if(!this.rates[date] || !this.rates[date][league] || !this.rates[date][league]['Currency'] || !this.rates[date][league]['Currency'][currency]) {
+      return 0;
+    }
     return this.rates[date][league]['Currency'][currency];
   }
 }

--- a/src/main/SettingsManager.ts
+++ b/src/main/SettingsManager.ts
@@ -3,6 +3,8 @@ import * as path from 'path';
 import * as fs from 'fs/promises';
 import { app } from 'electron';
 import DB from './db';
+import GGGAPI from './GGGAPI';
+import RateGetterV2 from './modules/RateGetterV2';
 
 const settingsPath = path.join(app.getPath('userData'), 'settings.json');
 
@@ -25,6 +27,14 @@ class SettingsManager {
     this.settings = require(path.join(app.getPath('userData'), 'settings.json'));
   }
 
+  async initializeDB(characterName: string) {
+    logger.info(`Initializing DB for ${characterName}`);
+    const character = (await GGGAPI.getAllCharacters()).find((character) => character.name === characterName);; 
+    await DB.initDB(character.name);
+    await DB.initLeagueDB(character.league, character.name);
+    await RateGetterV2.Getter.update();
+  }
+
   getAll() {
     return this.settings;
   }
@@ -40,10 +50,11 @@ class SettingsManager {
     if (
       key === 'activeProfile' &&
       value.characterName &&
-      this.settings.activeProfile &&
-      value.characterName !== this.settings.activeProfile.characterName
+      !!(this.settings.activeProfile || // First active Profile
+      (this.settings.activeProfile && // New active Profile
+      value.characterName !== this.settings.activeProfile.characterName))
     )
-      await DB.initDB(value.characterName);
+      await this.initializeDB(value.characterName);
     this.settings[key] = value;
     this.scheduleSave();
   }

--- a/src/main/db/index.ts
+++ b/src/main/db/index.ts
@@ -1,9 +1,7 @@
 import DatabaseConstructor, { Database, SqliteError } from 'better-sqlite3';
-// import { Database } from 'sqlite3';
 import * as path from 'path';
-// import logger from './Log'.getLogger(__filename);
-// import Utils from './Utils';
 import { get as getSettings } from './settings';
+import logger from 'electron-log';
 import { app } from 'electron';
 
 const sleep = (ms: number) => {
@@ -83,12 +81,9 @@ class DB {
       }
       char = settings.activeProfile.characterName;
     }
-    var app = require('electron').app || require('@electron/remote').app;
-    var db = new DatabaseConstructor(path.join(app.getPath('userData'), `${char}.db`));
+    const app = require('electron').app;
+    const db = new DatabaseConstructor(path.join(app.getPath('userData'), `${char}.db`), {verbose: logger.info});
     await this.init(db, initSQL, maintSQL);
-    // logger.info(`Completed initializing db for ${char}`);
-    // allow time for DB file changes to be written
-    await sleep(500);
     return db;
   }
 
@@ -135,70 +130,28 @@ class DB {
   }
 
   static async init(db: Database, sqlList: string[][], maintSqlList: string[] = []) {
-    // logger.info('Initializing ' + path.basename(db.filename));
-    return new Promise((resolve, reject) => {
-      let version = 0;
-      try {
-        version = db.pragma('user_version') as number;
-      } catch (err) {
-        // logger.info('Error reading database version: ' + err);
-        reject(err);
+    let version = 0;
+    try {
+      version = db.pragma('user_version', { simple: true }) as number;
+    } catch (err) {
+      logger.error('Error reading database version: ' + err);
+      return;
+    }
+
+    sqlList.forEach((commands, index) => {
+      if (version === 0 || index > version) {
+        commands.forEach((command) => {
+          logger.info(`Running initialization SQL for ${db.name} - version ${index}`);
+          db.prepare(command).run();
+        });
       }
-
-      sqlList.forEach((commands, index) => {
-        if (version === 0 || index > version) {
-          commands.forEach((command) => {
-            db.prepare(command).run();
-          });
-        }
-      });
-      maintSqlList.forEach((command) => {
-        db.prepare(command).run();
-      });
-
-      resolve(null);
-
-      // db.get('pragma user_version', (err, row) => {
-      //   if (err) {
-      //     // logger.info('Error reading database version: ' + err);
-      //     reject(err);
-      //   } else {
-      //     let ver = row.user_version;
-      //     // logger.info(`Database version is ${ver}`);
-      //     db.serialize(() => {
-      //       // for (let i = 0; i < sqlList.length; i++) {
-      //       //   if (ver === 0 || i > ver) {
-      //       //     // logger.info(`Running initialization SQL for version ${i}`);
-      //       //     for (let j = 0; j < sqlList[i].length; j++) {
-      //       //       let sql = sqlList[i][j];
-      //       //       // logger.info(sql);
-      //       //       db.run(sql, (err) => {
-      //       //         if (err) {
-      //       //           if (!err.toString().includes('duplicate column name')) {
-      //       //             // logger.info(`Error initializing DB: ${err}`);
-      //       //             reject(err);
-      //       //           }
-      //       //         }
-      //       //       });
-      //       //     }
-      //       //   }
-      //       // }
-      //       if (maintSqlList) {
-      //         for (let i = 0; i < maintSqlList.length; i++) {
-      //           db.run(maintSqlList[i], (err) => {
-      //             if (err) {
-      //               // logger.info(`Error running DB maintenance: ${err}`);
-      //               reject(err);
-      //             }
-      //           });
-      //         }
-      //         // logger.info('DB maintenance complete');
-      //       }
-      //       resolve();
-      //     });
-      //   }
-      // });
     });
+
+    maintSqlList.forEach((command) => {
+      db.prepare(command).run();
+    });
+
+    return null;
   }
 }
 
@@ -338,7 +291,7 @@ const initSQL = [
   // version 7 - properly set ignored tag in runinfo, instead of relying on magic numbers
   [
     'pragma user_version = 7',
-    `update mapruns set runinfo = json_set(ifnull(runinfo, "{}"), '$.ignored', true) where kills = -1 and gained = -1`,
+    `update mapruns set runinfo = JSON_SET(IFNULL(runinfo, '{}'), '$.ignored', true) where kills = -1 and gained = -1`,
   ],
 
   // version 8 - passive tree history
@@ -354,23 +307,23 @@ const initSQL = [
       update mapruns set runinfo = (
         select json_insert(
           runinfo, 
-          '$.masters."Einhar, Beastmaster".redBeasts', redbeasts, 
-          '$.masters."Einhar, Beastmaster".yellowBeasts', yellowbeasts
+          '$.masters.\`Einhar, Beastmaster\`.redBeasts', redbeasts, 
+          '$.masters.\`Einhar, Beastmaster\`.yellowBeasts', yellowbeasts
         ) as newinfo
         from (
-          select id, sum(case beast when "red" then 1 else 0 end) as redbeasts, sum(case beast when "yellow" then 1 else 0 end) as yellowbeasts from (
+          select id, sum(case beast when 'red' then 1 else 0 end) as redbeasts, sum(case beast when 'yellow' then 1 else 0 end) as yellowbeasts from (
             select case event_text
-              when "Einhar, Beastmaster: Haha! You are captured, stupid beast." then "yellow"
-              when "Einhar, Beastmaster: You have been captured, beast. You will be a survivor, or you will be food." then "yellow"
-              when "Einhar, Beastmaster: This one is captured. Einhar will take it." then "yellow"
-              when "Einhar, Beastmaster: Ohhh... That was a juicy one, exile." then "yellow"
-              when "Einhar, Beastmaster: Do not worry little beast! We are friends now!" then "yellow"
-              when "Einhar, Beastmaster: Off you go, little beast! Away!" then "yellow"
-              when "Einhar, Beastmaster: We will be best friends beast! Until we slaughter you!" then "yellow"
-              when "Einhar, Beastmaster: Great job, Exile! Einhar will take the captured beast to the Menagerie." then "red"
-              when "Einhar, Beastmaster: The First Ones look upon this capture with pride, Exile. You hunt well." then "red"
-              when "Einhar, Beastmaster: Survivor! You are well prepared for the end. This is a fine capture." then "red"
-              when "Einhar, Beastmaster: What? Do you not have nets, exile?" then "red"
+              when 'Einhar, Beastmaster: Haha! You are captured, stupid beast.' then 'yellow'
+              when 'Einhar, Beastmaster: You have been captured, beast. You will be a survivor, or you will be food.' then 'yellow'
+              when 'Einhar, Beastmaster: This one is captured. Einhar will take it.' then 'yellow'
+              when 'Einhar, Beastmaster: Ohhh... That was a juicy one, exile.' then 'yellow'
+              when 'Einhar, Beastmaster: Do not worry little beast! We are friends now!' then 'yellow'
+              when 'Einhar, Beastmaster: Off you go, little beast! Away!' then 'yellow'
+              when 'Einhar, Beastmaster: We will be best friends beast! Until we slaughter you!' then 'yellow'
+              when 'Einhar, Beastmaster: Great job, Exile! Einhar will take the captured beast to the Menagerie.' then 'red'
+              when 'Einhar, Beastmaster: The First Ones look upon this capture with pride, Exile. You hunt well.' then 'red'
+              when 'Einhar, Beastmaster: Survivor! You are well prepared for the end. This is a fine capture.' then 'red'
+              when 'Einhar, Beastmaster: What? Do you not have nets, exile?' then 'red'
               end 
             as beast from events
             where events.event_text like 'Einhar%'

--- a/src/main/index.ts
+++ b/src/main/index.ts
@@ -89,6 +89,7 @@ const init = async () => {
         league: character.league,
         valid: true,
       });
+      SettingsManager.initializeDB(character.name);
       logger.info(`Settings updated. Character: ${character.name}, League: ${character.league}`);
     } catch (e) {
       logger.error(

--- a/src/renderer/index.js
+++ b/src/renderer/index.js
@@ -12,6 +12,8 @@ import RunList from './routes/RunList';
 import Run from './routes/Run';
 import Login from './routes/Login';
 import Stats from './routes/Stats';
+import CharacterSelect from './routes/CharacterSelect';
+import LoginBox from './routes/LoginBox';
 import { electronService } from './electron.service';
 const { logger } = electronService;
 const runStore = new RunStore();
@@ -83,10 +85,24 @@ const router = createHashRouter([
   {
     path: '/login',
     element: <Login />,
-    loader: async () => {
-      const { code_challenge, state } = await ipcRenderer.invoke('oauth:get-info');
-      return { code_challenge, state };
-    },
+    children: [
+      {
+        index: true,
+        element: <LoginBox store={runStore} />,
+        loader: async () => {
+          const { code_challenge, state } = await ipcRenderer.invoke('oauth:get-info');
+          return { code_challenge, state };
+        },
+      },
+      {
+        path: 'character-select',
+        element: <CharacterSelect />,
+        loader: async () => {
+          const characters = await ipcRenderer.invoke('get-characters');
+          return { characters };
+        },
+      }
+    ]
   },
 ]);
 const darkTheme = createTheme({

--- a/src/renderer/routes/CharacterSelect.css
+++ b/src/renderer/routes/CharacterSelect.css
@@ -1,0 +1,19 @@
+.Character-Select {
+  display: flex;
+  flex-direction: column;
+  gap: 15px;
+  align-items: center;
+}
+.Character-Select__Form {
+  display: flex;
+  flex-direction: column;
+  gap: 15px;
+  align-items: center;
+}
+
+.Character-Select__Title {
+  font-family: FontinSmallCaps;
+  font-size: 1.2em;
+  font-weight: bold;
+  margin-bottom: 20px;
+}

--- a/src/renderer/routes/CharacterSelect.tsx
+++ b/src/renderer/routes/CharacterSelect.tsx
@@ -1,0 +1,69 @@
+import React from 'react';
+import FormControl from '@mui/material/FormControl';
+import Select, { SelectChangeEvent } from '@mui/material/Select';
+import FormHelperText from '@mui/material/FormHelperText';
+import MenuItem from '@mui/material/MenuItem';
+import Button from '@mui/material/Button';
+import Logo from '../assets/img/icons/png/128x128.png';
+import { useLoaderData, useNavigate } from 'react-router-dom';
+import './CharacterSelect.css';
+import { electronService } from '../electron.service';
+const { logger, ipcRenderer } = electronService;
+
+const CharacterSelect = () => {
+  const navigate = useNavigate();
+  const { characters } = useLoaderData() as any;
+  const [ character, setCharacter] = React.useState(characters.find((character) => character.current));
+  const [ isSaving, setIsSaving ] = React.useState(false);
+  const handleChange = (event: SelectChangeEvent<{ value: any }>) => {
+    setCharacter(characters.find((character) => character.name = event.target.value));
+  };
+
+  const saveCharacter = async () => {
+    logger.info('Saving active character from Login');
+    await setIsSaving(true);
+    const data = {
+      activeProfile: {
+        characterName: character.name,
+        league: character.league,
+        valid: true,
+      }
+    };
+  
+    await ipcRenderer.invoke('save-settings', { settings: data });
+    
+    navigate('/settings', { replace: true });
+  };
+
+  return (
+    <div className="Character-Select">
+      <img src={Logo} alt="Exile Diary Logo" className="Login__Logo" />
+      <div className='Character-Select__Title'>
+        Exile Diary <span className="Text--Legendary">Reborn</span>
+      </div>
+      <FormControl className="Character-Select__Form">
+        <FormControl size="small">
+          <Select id="character" value={character.name} onChange={handleChange}>
+            {characters.map((character: any) => (
+              <MenuItem key={character.name} value={character.name}>
+                {character.name}
+              </MenuItem>
+            ))}
+          </Select>
+          <FormHelperText>Pick a character to track</FormHelperText>
+        </FormControl>
+
+        <Button
+          variant="contained"
+          color="primary"
+          onClick={saveCharacter}
+          disabled={isSaving}
+          >
+          Validate
+        </Button>
+      </FormControl>
+    </div>
+  );
+};
+
+export default CharacterSelect;

--- a/src/renderer/routes/Login.tsx
+++ b/src/renderer/routes/Login.tsx
@@ -1,84 +1,11 @@
-import React, { useEffect, useState } from 'react';
-import { Button } from '@mui/material';
-import { useParams, useLoaderData, useNavigate } from 'react-router';
-import { electronService } from '../electron.service';
-import Logo from '../assets/img/icons/png/128x128.png';
+import React from 'react';
+import { Outlet } from 'react-router-dom';
 import './Login.css';
-import { ipcRenderer } from 'electron';
-const { openExternal } = electronService.shell;
-const { logger } = electronService;
 
-type AuthData = {
-  state: string;
-  code_challenge: string;
-};
-
-function Login({}) {
-  const { code_challenge, state } = useLoaderData() as AuthData;
-  const url =
-    'https://www.pathofexile.com/oauth/authorize?' +
-    'client_id=exilediaryreborn' +
-    '&response_type=code' +
-    '&scope=account:characters account:stashes account:league_accounts account:item_filter' +
-    `&state=${state}` +
-    '&redirect_uri=https://exilediary.com/auth/success' +
-    `&code_challenge=${code_challenge}` +
-    '&code_challenge_method=S256';
-
-  const openLink = () => {
-    openExternal(url);
-    setIsError(false);
-    setIsOngoing(true);
-  };
-  const navigate = useNavigate();
-  const [isFetchingOauthToken, setIsFetchingOauthToken] = useState(false);
-  const [isOngoing, setIsOngoing] = useState(false);
-  const [isError, setIsError] = useState(false);
-
-  useEffect(() => {
-    ipcRenderer.on('oauth:auth-failure', (event, arg) => {
-      logger.info('Auth Failure, redirecting to the root page');
-      setIsOngoing(false);
-      setIsFetchingOauthToken(false);
-      setIsError(true);
-    });
-
-    ipcRenderer.on('oauth:received-code', (event, arg) => {
-      setIsFetchingOauthToken(true);
-    });
-
-    ipcRenderer.on('oauth:auth-success', (event, arg) => {
-      logger.info('Auth Success, redirecting to the root page');
-      navigate('/', { replace: true });
-    });
-
-    setIsError(false);
-    setIsOngoing(false);
-    setIsFetchingOauthToken(false);
-  }, []);
-
-  const warning =
-    isOngoing && !isFetchingOauthToken ? (
-      <p>Please authenticate through the window that just opened</p>
-    ) : null;
-
+const Login = () => {
   return (
     <div className="Login">
-      <div className="Login__Box">
-        <img src={Logo} alt="Exile Diary Logo" className="Login__Logo" />
-        <h3>
-          Exile Diary <span className="Text--Legendary">Reborn</span> requires you to log in with
-          the PoE API to function
-        </h3>
-        {isError ? <p className="Test--Error">Something went wrong, please try again</p> : null}
-        <Button variant={isOngoing ? 'outlined' : 'contained'} color="primary" onClick={openLink}>
-          Login with PoE
-        </Button>
-        {isFetchingOauthToken ? <p>Received Code, Fetching Oauth Token...</p> : null}
-        {isOngoing && !isFetchingOauthToken ? (
-          <p>Please authenticate through the window that just opened</p>
-        ) : null}
-      </div>
+      <Outlet />
     </div>
   );
 }

--- a/src/renderer/routes/LoginBox.tsx
+++ b/src/renderer/routes/LoginBox.tsx
@@ -1,0 +1,84 @@
+import React, { useEffect, useState } from 'react';
+import { Button } from '@mui/material';
+import { useLoaderData, useNavigate, Outlet } from 'react-router';
+import { electronService } from '../electron.service';
+import Logo from '../assets/img/icons/png/128x128.png';
+import './Login.css';
+import { ipcRenderer } from 'electron';
+const { openExternal } = electronService.shell;
+const { logger } = electronService;
+
+type AuthData = {
+  state: string;
+  code_challenge: string;
+};
+
+const LoginBox = ({}) => {
+  const { code_challenge, state } = useLoaderData() as AuthData;
+  const url =
+    'https://www.pathofexile.com/oauth/authorize?' +
+    'client_id=exilediaryreborn' +
+    '&response_type=code' +
+    '&scope=account:characters account:stashes account:league_accounts account:item_filter' +
+    `&state=${state}` +
+    '&redirect_uri=https://exilediary.com/auth/success' +
+    `&code_challenge=${code_challenge}` +
+    '&code_challenge_method=S256';
+
+  const openLink = () => {
+    openExternal(url);
+    setIsError(false);
+    setIsOngoing(true);
+  };
+  const navigate = useNavigate();
+  const [isFetchingOauthToken, setIsFetchingOauthToken] = useState(false);
+  const [isOngoing, setIsOngoing] = useState(false);
+  const [isError, setIsError] = useState(false);
+
+  useEffect(() => {
+    ipcRenderer.on('oauth:auth-failure', (event, arg) => {
+      logger.info('Auth Failure, redirecting to the root page');
+      setIsOngoing(false);
+      setIsFetchingOauthToken(false);
+      setIsError(true);
+    });
+
+    ipcRenderer.on('oauth:received-code', (event, arg) => {
+      setIsFetchingOauthToken(true);
+    });
+
+    ipcRenderer.on('oauth:auth-success', (event, arg) => {
+      logger.info('Auth Success, redirecting to the root page');
+      navigate('/login/character-select', { replace: true });
+    });
+
+    setIsError(false);
+    setIsOngoing(false);
+    setIsFetchingOauthToken(false);
+  }, []);
+
+  const warning =
+    isOngoing && !isFetchingOauthToken ? (
+      <p>Please authenticate through the window that just opened</p>
+    ) : null;
+
+  return (
+    <div className="Login__Box">
+      <img src={Logo} alt="Exile Diary Logo" className="Login__Logo" />
+      <h3>
+        Exile Diary <span className="Text--Legendary">Reborn</span> requires you to log in with
+        the PoE API to function
+      </h3>
+      {isError ? <p className="Test--Error">Something went wrong, please try again</p> : null}
+      <Button variant={isOngoing ? 'outlined' : 'contained'} color="primary" onClick={openLink}>
+        Login with PoE
+      </Button>
+      {isFetchingOauthToken ? <p>Received Code, Fetching Oauth Token...</p> : null}
+      {isOngoing && !isFetchingOauthToken ? (
+        <p>Please authenticate through the window that just opened</p>
+      ) : null}
+    </div>
+  );
+}
+
+export default LoginBox;

--- a/src/renderer/routes/Settings.tsx
+++ b/src/renderer/routes/Settings.tsx
@@ -42,7 +42,8 @@ function a11yProps(index: number) {
 const Settings = () => {
   const navigate = useNavigate();
   const { settings, characters } = useLoaderData() as SettingsLoaderData;
-  const [tabValue, setTabValue] = React.useState(0);
+  const [ tabValue, setTabValue ] = React.useState(0);
+  const [ errors, setErrors ] = React.useState<any>({});
 
   // Character
   const [character, setCharacter] = React.useState(
@@ -94,8 +95,24 @@ const Settings = () => {
     setTabValue(newValue);
   };
 
+  const handleClientLocationPick = (e) => {
+    if(!e.target.value) {
+      setErrors({ ...errors, clientFileLocation: 'POE Client file location is required' });
+      return;
+    } else {
+      setErrors({ ...errors, clientFileLocation: null });
+      setClientFileLocation(e.target.value);
+    }
+  };
+
   const handleSubmit = (e) => {
     e.preventDefault();
+    if(!clientFileLocation) {
+      setErrors({ ...errors, clientFileLocation: 'POE Client file location is required' });
+      return;
+    } else {
+      setErrors({ ...errors, clientFileLocation: null });
+    }
     const data = {
       activeProfile: {
         characterName: character,
@@ -175,7 +192,9 @@ const Settings = () => {
                 variant="filled"
                 size="small"
                 value={clientFileLocation}
-                onChange={(e) => setClientFileLocation(e.target.value)}
+                helperText={errors.clientFileLocation ? errors.clientFileLocation : ''}
+                error={!!errors.clientFileLocation}
+                onChange={handleClientLocationPick}
               />
               <Button component="label">
                 Choose Folder

